### ISSUE TITLE
Network - does updating an edge cause it to reset automatically in Hierarchical view? 

### ISSUE THREAD

**Issue by dockstreet** [almende/vis#3150]
Hi,

For VisJS 4.20.0 & 4.19.0 - is updating an edge problematic? I seem to be having issue through a wrapper but can't tell?

https://github.com/datastorm-open/visNetwork/issues/162

thx

**Comment by wimrijnders** [comment]
Hi there, my knowledge of R is non-existent, so I really can't say anything worthwhile about the code in the referenced issue. So I guess this needs to be looked at from a higher level.

I have never heard of any issues as described concerning the updating of edges. Would you mind describing a test case with aberrant behavior? I will then try it in javascript and see if I can reproduce it.

**Comment by dockstreet** [comment]
Will do - could take me a bit of time, I'll be travelling over the next week - and I do appreciate the help . 

It's a strange combination of hierarchical layout & edge label resizing in an event - it for some reason causes a reset.

**Comment by dockstreet** [comment]
Hi, I've been able to recreate the issue in JavaScript - sorry for the delayed response. 

It seems if hidden is not set to true/false it will not reset as one would expect. I actually need to set it to hidden at some points in to make the code dynamic. 

http://jsfiddle.net/dockstreet/mxzjf7j8/

Let me know if you need any more information to reproduce, if you drag node 3 over then hover over an edge - it will reset unexpectedly.

**Comment by wimrijnders** [comment]
Thanks for the demo, this sort of looks familiar. Resets of initial coordinates is a recurring issue for non-hierarchical networks, there seems to be a relation.

> (by the way, this library is the best network library out there, by leaps and bounds!!!)

❤️ Nice to  be appreciated.

**Comment by wimrijnders** [comment]
Confirmed.

OK, so this is about the explicit setting of the `hidden` option per node. I can confirm it has an effect; `Edge.setOptions(), line 100:`

```js
    if (options.hidden !== undefined || options.physics !== undefined) {
      dataChanged = true;
    }
```

So, regardless of its true/false value, the node will be flagged as 'changed'. The consequence of this is that the entire layout will be recalculated, leading to the jump you saw in the view.

I'm going to spend a little time contemplating why this is so and what can be done about it. I don't have a workaround for you right now, perhaps I'll think about it. The easiest of course is to not use the `hidden` option, but you indicated that you need it.

**Comment by wimrijnders** [comment]
How's this for a workaround? Remove the `hidden` options from the edges and do this:

```js
network.on('hoverEdge', function(e) {
    this.body.data.edges.update({
        id: e.edge,
        font: {
            size: 14
        },
        color: {
          opacity: 1
        }
    });
});

network.on('blurEdge', function(e) {
    this.body.data.edges.update({
        id: e.edge,
        font: {
            size: 0
        },
        color: {
          opacity: 0
        }
    });
});
```

I.e. make the edges transparent. This won't trigger a recalculation of the layout.

**Comment by dockstreet** [comment]
It does help a lot!!! It does solve the problem of the edges resetting the nodes as a workaround. 

Following this guidance, I do have one issue however not with hiding edges, but nodes. Seems like for my use case it causes a reset when because I have to use hidden on a node to try to match the opacity for an edge.

So I think the opacity if implemented for a node without a recalculation of the layout would help.

In fact... I think you are already looking into this https://github.com/almende/vis/issues/1185

**Comment by wimrijnders** [comment]
Yeah, #1185 will be implemented eventually, but it's not scheduled yet. I hope you are not holding your breath till then.

**Comment by dockstreet** [comment]
Any other workarounds / thoughts for hiding the nodes that would not cause a recalculation of the layout? 

I tried setting the color, I guess I could make it like the background but the problem is if someone tries to drag it then it would overlay other nodes.

**Comment by wimrijnders** [comment]
CSS doesn't work on nodes. `Network` draws on a canvas and there is no relation with css and DOM. It's all graphical.

`z-index` is not implemented yest, but it's on the feature list, see #3146 (I see you found it yourself). You can sort of manipulate it, because the nodes and edges are drawn in the order that they are added to the `DataSet`'s. It's a bit nifty, though.

In short, it's not possible right now to hide nodes/edges without resetting, or make invisible nodes unclickable. This is going to need an API change to fix it. I'm thinking of making the reset of the network optional, or said in another way, add an option to stop the reset. There's no workaround for this right now, sorry.

**Comment by dockstreet** [comment]
Thanks! - looking forward to the API changes to fix/address it - much appreciated for looking into this

**Comment by dockstreet** [comment]
Would adding the flag to stop the reset perhaps make it into the next version? That would really help
again thanks! - looking forward to the API changes to fix/address it - much appreciated for looking into this

Question - would saving the x,y positions of the nodes and edges be possible without the API changes ? If so, how could we save our position ?

**Comment by wimrijnders** [comment]
@dockstreet There are no plans for adding the flag right now. However, there is also no date set for the next release, so it's still possible.

As to saving/restoring the positions in hierarchical networks, I believe that is actually not possible right now. `getPositions()` returns the coordinates, as you point out, but I don't think there's a way to get the actual levels. I'll need to look into this.

**Comment by dockstreet** [comment]
@wimrijnders would be great to see the flag added to the next release, could definitely use it - if possible.  

Thanks for looking into this !!!!

**Comment by wimrijnders** [comment]
I'm busy with other more pressing issues right now, it's still on my TODO.
Hereby noting that the resetting of options is a duplicate of #1660.

Also, z-indexing is a duplicate of #3146. I'm inclined to cloe this issue to continue with the mentioned ones. Is that OK with you?

**Comment by dockstreet** [comment]
Glad it is still on your TODO - thx

I'll keep following this issue, and would be glad to test when an update has been made - kind regards

**Comment by dockstreet** [comment]
just a polite bump on this to keep it active, I still have interest - would be super if flag make it into next release - cheers!

**Comment by wimrijnders** [comment]
Bump is noted. I'm just not getting around to it, sorry!
